### ISSUE TITLE
Whitelist common false-positives

### ISSUE THREAD

**Issue by veloute** [hectorm/hblock#20]
It might be a good idea to add a list of commonly whitelisted domains such as [this](https://raw.githubusercontent.com/anudeepND/whitelist/master/domains/whitelist.txt) to lower the probability of people having problems.

Found the list at https://firebog.net/ under "Whitelisting Suggestions".

**Comment by hectorm** [comment]
I prefer to solve false positives at the root and contact the maintainer of the original list, I have done it in some occasions and it has worked. But I understand that it is a tedious operation and not always the maintainer will eliminate that entry from the list.

Currently the "--whitelist" argument of my script requires a list of regular expressions, so it does not directly support the whitelist you suggest. I will think about how to implement this feature without breaking the compatibility.

Meanwhile, I checked the domains that are currently present in both hBlock and anudeepND whitelist and these are the results.

```bash
comm -12 <(curl -fsS https://hblock.molinero.xyz/hosts_domains.txt | sort) <(curl -fsS https://raw.githubusercontent.com/anudeepND/whitelist/master/domains/whitelist.txt | sort)
```

| Domain                                                                                                                    | Blocked by                                                                    |
|:-------------------------------------------------------------------------------------------------------------------------:|:-----------------------------------------------------------------------------:|
| [cdn3.optimizely.com](https://github.com/anudeepND/whitelist/commit/da84f2d62fe6bbeda042d1088e1a773d72154c3a)             | blocklist.kowabit.de, winhelp2002.mvps.org                                    |
| [cdn.optimizely.com](https://github.com/anudeepND/whitelist/commit/d7e8c8bad2a5f6cb9893fd04e4dd939dc7ea4bf4)              | winhelp2002.mvps.org                                                          |
| [d2c8v52ll5s99u.cloudfront.net](https://github.com/anudeepND/whitelist/commit/4b34b67654bc3c1f92ce6afeeeb8541798a90d89)   | winhelp2002.mvps.org                                                          |
| [nexusrules.officeapps.live.com](https://github.com/anudeepND/whitelist/commit/1508153254958b8bb9c998f8417d625b7dc17166)  | stevenblack                                                                   |
| [om.cbsi.com](https://github.com/anudeepND/whitelist/commit/fced35868d45f3e2d46fd514c39dde036de5c394)                     | adguard-simplified, blocklist.kowabit.de, easyprivacy, winhelp2002.mvps.org   |
| [s.shopify.com](https://github.com/anudeepND/whitelist/commit/a4e5539786d40197610d5afea990fa60b04e0073)                   | blocklist.kowabit.de, winhelp2002.mvps.org                                    |
| [s.youtube.com](https://github.com/anudeepND/whitelist/commit/d7e8c8bad2a5f6cb9893fd04e4dd939dc7ea4bf4)                   | blocklist.kowabit.de, someonewhocares.org                                     |
| [v.shopify.com](https://github.com/anudeepND/whitelist/commit/a4e5539786d40197610d5afea990fa60b04e0073)                   | adguard-simplified                                                            |

**Comment by veloute** [comment]
I edited the script to include the list like this:

`whitelist="$(curl https://raw.githubusercontent.com/anudeepND/whitelist/master/domains/whitelist.txt)"`

it seems to work fine without any problems.

**Comment by hectorm** [comment]
Even though it seems to work, you are actually whitelisting more domains than you think, because the `--whitelist` option uses regular expressions instead of literal strings.

If you want to modify the script, I suggest you do this instead:
```sh
whitelist=$(curl -fsSL 'https://raw.githubusercontent.com/anudeepND/whitelist/master/domains/whitelist.txt' \
	| while read -r d; do printf '^%s$\n' "$(quoteRe "$d")"; done \
)
```

**Comment by veloute** [comment]
Huh, you're right. I don't have a tonne of experience with bash scripting, so thanks for that.

Would there be any negative reason to add this by default? You could even mirror the whitelist or host your own to only include domains which shouldn't be blocked but are in the blocklists.

**Comment by hectorm** [comment]
The solution I have given in https://github.com/hectorm/hBlock/issues/20#issuecomment-405306500 works for this case, but it is not generic enough, since it assumes that each line is a domain name, and is also too slow. In order to implement this feature I should first sanitize the input and improve the performance.

I still have improvements to make to the `--blacklist` and `--whitelist` options, so I will leave this issue pending until then.

On the other hand, are any of the domains that I mentioned in https://github.com/hectorm/hBlock/issues/20#issuecomment-405084908 causing you problems?

**Comment by erikdubois** [comment]
Can I thank you first for making this script - promoted it here - 
https://arcolinux.com/use-hblock-to-improve-your-security-and-privacy-by-blocking-ads-tracking-and-malware-domains/

Can I add something to the wishing list?

What if you provide a .config/hblock/whitelist.txt to the users so they can add an url rather then a regex (that is not everybody's cup of tea).

Currently I want to whitelist analytics.google.com to see my traffic. Regex works fine but we have implemented your application in a timer and a service. 

At that point the whitelisting is not applied or gone again and I need to manually intervene to whitelist it. 
IF we could have a file with either regex or url it would be a great improvement to solve the false positives.

**Comment by hectorm** [comment]
I'm thinking of making the `--whitelist` and `--blacklist` options support only files, disable by default the regex support in the whitelist and add a `-r, --whitelist-regex` option to enable it.

Clearly this breaks compatibility, so I must increase the version to 2.0.0.

**Comment by erikdubois** [comment]
so --whitelist-regex will only support regex and no url and  will read a file that contains 1 specific rule on each line?

and

whitelist will only support a file with urls in it and will read a file that contains 1 specific url on each line?

or is it something else?

Since Hblock runs by default in our system... it is best we supply already url's that we whitelist for the users like google analytics.

So it would be great that if you just type hblock the application looks for the two files and already includes or excludes what is in those files.If no files or content ... nothing happens.

So it is up to the users to manage their own whitelist and blacklist.

**Comment by hectorm** [comment]
I have created an [experimental branch](https://github.com/hectorm/hblock/tree/whitelist-and-blacklist-improvements) with these changes and other improvements. I haven't updated the readme, but you can use the `--help` option to understand how to use it.

**Comment by erikdubois** [comment]
Thanks for your effort.
Tried it out and looks promising in code... does not work yet.
I am sure it will work. 
Here are two screenshots of what I tried.
![arcolinux-2018-11-23-1542985487_screenshot_3840x1080](https://user-images.githubusercontent.com/10594806/48949880-cd386680-ef39-11e8-900c-3afa93e79874.jpg)
![arcolinux-2018-11-23-1542985478_screenshot_3840x1080](https://user-images.githubusercontent.com/10594806/48949881-cd386680-ef39-11e8-9162-172fd0864961.jpg)

**Comment by hectorm** [comment]
The `whitelist.list` file must contain domain names separated by new lines. So instead of `https://analytics.google.com` it should be `analytics.google.com`.

**Comment by erikdubois** [comment]
it worked with this line of code
`./hblock --whitelist /etc/hblock.d/whitelist.list` using analytics.google.com in the whitelist list file
Could we program that IF there is nothing behind --whitelist we fall back on the /etc/hblock.d/whitelist.list file if it is there.

so 

`./hblock --whitelist`

would be become the same thing or is that what this variable is for?

**Comment by hectorm** [comment]
In the new branch the whitelist works as follows.

If the environment variable `HBLOCK_WHITELIST` is defined, its value is used as whitelist, otherwise it checks the existence of the file `/etc/hblock.d/whitelist.list` and its content is used as whitelist, otherwise it uses the value built in the script itself, which is currently empty.

With the option `--whitelist` you can specify another file different from `/etc/hblock.d/whitelist.list`. Although thanks to your comments you have made me realize that if the user explicitly uses the `--whitelist` option, it should have priority over the environment variable `HBLOCK_WHITELIST`.

**Comment by erikdubois** [comment]
looking good - lets move forward
The installation of hblock should create the /etc/hblock.d/whitelist.list and blacklist.list automatically ;-)

**Comment by erikdubois** [comment]
Ok If you want me to test it before release...
I will make the time.

Op za 24 nov. 2018 om 16:26 schreef Héctor Molinero Fernández <
notifications@github.com>:

**Comment by hectorm** [comment]
I'd really appreciate it if you'd take the time to check it out. Thank you!

**Comment by hectorm** [comment]
Done (d2f935aace03acd74bbfbe43588c5a7cc5e1eebd).

**Comment by erikdubois** [comment]
super

Op zo 25 nov. 2018 om 13:36 schreef Héctor Molinero Fernández <
notifications@github.com>:

**Comment by erikdubois** [comment]
and how to test - stay in the experimental branch?

Op zo 25 nov. 2018 om 13:36 schreef Héctor Molinero Fernández <
notifications@github.com>:

**Comment by erikdubois** [comment]
Crazy idea... this is what I did without reading the posts again.
created a folder in ~/.config/hblock and put whitelist file in it...
then read it again
This way it is in the users directory with full read and write capabilities...
![arcolinux-2018-11-29-1543484439_screenshot_3840x1080](https://user-images.githubusercontent.com/10594806/49212907-685e8f80-f3c3-11e8-92ce-85491b48e4cf.jpg)

**Comment by erikdubois** [comment]
tried the experimental branch and typing hblock works
/etc/hblock.d/whitelist.list containing google.analytics.com
Just for your information.
Here is our PKGBUILD
We will need to adapt to add a folder with the whitelist.list containing ... false positives
The question is where shall hblock read its list.
/etc/hblock.d/whitelist.list
or
~/.config/hblock/whitelist.list

Should we not add already a file blacklist.list to the pkgbuild?

**Comment by hectorm** [comment]
I think the correct path should be `/etc/hblock.d/whitelist.list` as the user needs permission to modify the `/etc/hosts` file anyway and it is a change that affects all users.

On the other hand, since you are including the `/etc/hblock.d/whitelist.list` file, I think it's a good idea to also include `/etc/hblock.d/blacklist.list` for consistency. I have no plans to use any blacklist by default.

**Comment by erikdubois** [comment]
as I said it is up to you...
consistency is key- agreed

Op do 29 nov. 2018 om 19:20 schreef Héctor Molinero Fernández <
notifications@github.com>:

**Comment by erikdubois** [comment]
let me know when I can test again
cheers

Op do 29 nov. 2018 om 19:20 schreef Héctor Molinero Fernández <
notifications@github.com>:

**Comment by hectorm** [comment]
All right, I just released version 2.0.0.

**Comment by erikdubois** [comment]
works fine
